### PR TITLE
support non-cell-wise owned state

### LINKED ISSUE #422: C04SoA traversal produces wrong results when using Halo particles
**Describe the bug**
The C04SoA Traversal produces wrong results when using Halo particles.

**Caused by**
The C04SoA traversal combines owned and halo particles into one SoA.
It calls the LJ SoA functor with these combined SoAs, it will calculate wrong global values, because it assumes one owned state for the entire combined SoA.

**Proposed fix**
Support for non-cell-wise owned state should be added. It should be needed for VerletClusterCells, anyways.

**Minimal example**
Replace particle addition in TraversalComparison (boxmax: {3.,3.,3.}, cellsizefactor=1.) with:
```c++
size_t id = 0;
container->addParticle(autopas::MoleculeLJ({1.86071, 1.54458, 1.06225}, {0., 0., 0.}, id++));
container->addParticle(autopas::MoleculeLJ({1.06143, 2.8866, 0.478508}, {0., 0., 0.}, id++));
container->addHaloParticle(autopas::MoleculeLJ({1.58559, 3.55956, 0.74583}, {0., 0., 0.}, id++));
```
and set appropriate particle counts in `TraversalComparison::getTestParams()`

**To Reproduce**
1. Enable C04SoATraversal again (currently completely disabled!)
2. Run TraversalComparison test.

OR use the above minimal example.

**Expected behavior**
Correct calculation of globals.

### PR DESCRIPTION
**Pull request by SteffenSeckler**
# Description

Aims to fix #422 by adding support for non-cell-wise owned state.

- [x] LJFunctor
   - [x] SoAFunctorPair
   - [x] SoAFunctorSingle
- [x] LJFunctorAVX
   - [x] SoAFunctorPair
   - [x] SoAFunctorSingle
- [x] SPHCalcDensityFunctor -- invalid, not needed
- [x] SPHHydroForceFunctor -- invalid, not needed

## TODO
- [x] Reenable C04SoATraversal
- [x] Change number in AutoTunerTest::testAllConfigurations

## Related Pull Requests

- #PR

## Resolved Issues

- [x] fixes #422 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] should be tested with tests from #301 
- [x] Add tests for different functors for both cellWiseOwnedState true and false
   - [x] LJFunctor
   - [x] LJFunctorAVX

### CODE REVIEW

File: src/autopas/containers/linkedCells/traversals/C04SoACellHandler.h [hunk 1]
```diff
@@ -211,13 +211,13 @@ inline void C04SoACellHandler<ParticleCell, PairwiseFunctor, dataLayout, useNewt
         auto stripeView = cell1->_particleSoABuffer.constructView(0, numParticlesBaseCell);
         if (slice == currentSlice) {
           // process stripe with itself
-          _pairwiseFunctor->SoAFunctor(stripeView, useNewton3);
+          _pairwiseFunctor->SoAFunctorSingle(stripeView, useNewton3);
```
Comment (SteffenSeckler) [inline src/autopas/containers/linkedCells/traversals/C04SoACellHandler.h]: @HobbyProgrammer does the C04SoATraversal pass a view to the SoAFunctor that potentially has owned AND non-owned particles?

Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/traversals/C04SoACellHandler.h]: also, check the other calls to SoAFunctorPair, please :)

Comment (HobbyProgrammer) [reply src/autopas/containers/linkedCells/traversals/C04SoACellHandler.h]: @SteffenSeckler I don't know, I have never looked deeper into how this traversal works. It was just ported to `SoAView`s from a previous implementation of views in the `SoA`. @Garfield96 maybe knows!

Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/traversals/C04SoACellHandler.h]: push @Garfield96

Comment (Garfield96) [reply src/autopas/containers/linkedCells/traversals/C04SoACellHandler.h]: @SteffenSeckler Potentially yes. This traversal does not distinguish between owned and non-owned (at least it didn't then I implemented it)

Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/traversals/C04SoACellHandler.h]: ok. thanks for the info.

**Comment by SteffenSeckler** [comment]
@FG-TUM you may already look at the avx functor for this ;)

**Comment by SteffenSeckler** [comment]
the rest is more or less done, just thinking of refactoring the LJFunctorTest to generated tests.

**Comment by SteffenSeckler** [comment]
@FG-TUM fix for LJFunctorAVX is up to you, now ;)

File: src/autopas/molecularDynamics/LJFunctor.h [hunk 2]
```diff
@@ -214,9 +215,10 @@ class LJFunctor
     SoAFloatPrecision shift6 = _shift6;
     SoAFloatPrecision sigmasquare = _sigmasquare;
     SoAFloatPrecision epsilon24 = _epsilon24;
-    if (calculateGlobals) {
-      // Checks if the cell is a halo cell, if it is, we skip it.
+    const bool duplicatedCalculations = _duplicatedCalculations;
+    if (calculateGlobals and cellWiseOwnedState and _duplicatedCalculations) {
```
Comment (FG-TUM) [inline src/autopas/molecularDynamics/LJFunctor.h]: why copying this and then not using it?
Something about autovec?

**Comment by FG-TUM** [comment]
@SteffenSeckler Instead of duplicating all the tests for LJFunctor and LJFunctorAVX we should write them only once and Type-Parameterize them.

**Pull request merged**